#CU-2nvr7n3 governance and upgrades process description (#1386)
ComposableFi/composable: rfcs/0006-picasso-runtime-upgrades.md

# Overview

Proposes flow to do runtime upgrades of Picasso. 

> In its core [upgrades] are as described by Parity for developer network.

Reviewer: ```suggestion
In its core [upgrades]: https://docs.substrate.io/tutorials/get-started/forkless-upgrade are as described by Parity for the developer network.
```

Extended with security measures, like Karura and Kusama have, to protect liqiudity on these networks.

> Expected that the reader has understood official Parity documents and [explainers], [upgrades], [governance].
> 
> It is helpfull if reader undrestans *nix [substitute user] operation. 

Reviewer: ```suggestion
It is expected that the reader has understood: 
- [Official Parity documents](https://docs.substrate.io/)
- [explainers](https://www.youtube.com/playlist?list=PLOyWqupZ-WGuAuS00rK-pebTMAOxW41W8)
- [upgrades](https://docs.substrate.io/tutorials/get-started/forkless-upgrade)
- [governance](../doc/governance.md)

It is helpful if the reader understands *nix [substitute user](https://en.wikipedia.org/wiki/Sudo) operation. 
```

> Document is not intentended to described detailed specificaitons of quality gates.
> 
> Document specifies relevant domains and stakeholders and teams, but not names exactly what communication channel these would use as communicaiton medium.

Reviewer: ```suggestion
This document is not intentended to described detailed specificaitons of quality gates.

The Document specifies relevant domains and stakeholders and teams, but doesn't name exactly what communication channel these would use as communication medium.
```

## Constraints

- Upgrades can be done only under `Root` origin, but cannot be `single signature`. And yet we must be able to release very fast, for examples in case of hotfix.

- Upgrades are not immediate because runtime should be copied to parachain collators and relay validators and be enabled simultaneously on a specific block in the future.

- There should be aligment of stakeholders on quality of release to be sure that bridged Picasso assets are secure. 

## Legend

***represent on chain things*** which can be interacted via Polkadot.js or Composable parachain SDK.

## Flow

> ### Overview of flow
> 
> Relase notes collected, specifically `git commit`s which must be inlcuded into release are traceable to projects or pull request `item`s.
> 
> Relevant notes and references to runtime build artifact (`wasm`) is shared into company channel with relevant representativses.
> 
> Represantatives agree with this release.
> 
> Release is stored as ***preimage*** on Picasso. Preimage `hash` is shared in channel with ***counci*** representatives.
> 
> Council creates default YES ***motion*** to vote on enactment of premiage on democracy. Shareholdes and ***technicalCollective*** fast track enactment of new runtime.
> 
> Each step executed Here is mapping of teams to roles they play in process

Reviewer: ```suggestion
### Overview of flow

Release notes collected, specifically `git commits` which must be included in the release, are traceable to projects or pull request `items`.

Relevant notes and references to runtime build artifact (`wasm`) are shared in the company channel with relevant representatives.

Representatives agree with this release.

The release is stored as ***preimage*** on Picasso. Preimage `hash` is shared in a channel with ***council*** representatives.

Council creates default YES ***motion*** to vote on the enactment of preimage on democracy. Shareholders and ***technicalCollective*** fast track enactment of new runtime.

Each step executed here is mapping teams to roles they play in the process.
```

> #### Overview of actors
> 
> | role | groups of people| area |  
> |-----------------|-----|-----|
> | release engineers | `@ComposableFi/sre`| | 
> | integration stakeholders| not direct contributers | depend on runtime UI/Frontend/Bots/Explorer/Data integation |
> | council |   ***concil***(on chain list of keys with attahed ***identity***) | tokenomics, marketing |
> | technical| ***technicalCollective*** | speed up on chain changes | 

Reviewer: ```suggestion
#### Overview of actors

| role | groups of people| area |  
|-----------------|-----|-----|
| release engineers | `@ComposableFi/sre`| | 
| integration stakeholders| not direct contributors | depend on runtime UI/Frontend/Bots/Explorer/Data integration |
| council |   ***council***(on chain list of keys with attached ***identity***) | tokenomics, marketing |
| technical| ***technicalCollective*** | speed up on chain changes | 
```

> #### Overview of actors
> 
> | role | groups of people| area |  
> |-----------------|-----|-----|
> | release engineers | `@ComposableFi/sre`| | 
> | integration stakeholders| not direct contributers | depend on runtime UI/Frontend/Bots/Explorer/Data integation |
> | council |   ***concil***(on chain list of keys with attahed ***identity***) | tokenomics, marketing |
> | technical| ***technicalCollective*** | speed up on chain changes | 

Reviewer: ```suggestion
#### Overview of actors

| role | groups of people| area |  
|-----------------|-----|-----|
| release engineers | `@ComposableFi/sre`| | 
| integration stakeholders| not direct contributors | depend on runtime UI/Frontend/Bots/Explorer/Data integration |
| council |   ***council***(on chain list of keys with attached ***identity***) | tokenomics, marketing |
| technical| ***technicalCollective*** | speed up on chain changes | 
```

> #### Overview of core quality gates 
> 
> Minimal set of gates.
> 
> All changes to runtime are forward compatible.
> Specifically, if extinsic was added, it is retained and new version are added with larger dispatch identifiers.
> If storage was added, it is migrated by relevant runtime code on upgrade to new version. Storage is not ugprade in place as there consumers which may not be able to read it after it is changed. Same relates to events.
> 
> Runtime was upraded on Dali Rococo Testnet and was producing blocks here at for several hours.
> 
> Commit form which release is done passed full set of relevant check on protected main branch. 
> You cannot force push. 
> Each protocol on chain has ability to stop operating as of decision of its governance without dependency on runtime upgrade.
> 
> Resonable represenatives of stakeholders approved release to Piccaso.

Reviewer: ```suggestion
#### Overview of core quality gates 

Minimal set of gates.

All changes to runtime are forward compatible.
Specifically, if extrinsic was added, it is retained, and new versions are added with larger dispatch identifiers.
If storage was added, it is migrated by relevant runtime code on upgrade to the new version. Storage is not upgraded in place as consumers may not be able to read it after it is changed. The same relates to events.

Runtime was upgraded on Dali Rococo Testnet and produced blocks for several hours.

Commit form that confirms release is done, passed the full set of relevant checks on the protected main branch. 
You cannot force push. 
Each protocol on the chain can stop operating as a decision of its governance without dependency on runtime upgrade.

Reasonable representatives of stakeholders approved the release to Piccaso.
```

> #### Overview of core quality gates 
> 
> Minimal set of gates.
> 
> All changes to runtime are forward compatible.
> Specifically, if extinsic was added, it is retained and new version are added with larger dispatch identifiers.
> If storage was added, it is migrated by relevant runtime code on upgrade to new version. Storage is not ugprade in place as there consumers which may not be able to read it after it is changed. Same relates to events.
> 
> Runtime was upraded on Dali Rococo Testnet and was producing blocks here at for several hours.
> 
> Commit form which release is done passed full set of relevant check on protected main branch. 
> You cannot force push. 
> Each protocol on chain has ability to stop operating as of decision of its governance without dependency on runtime upgrade.
> 
> Resonable represenatives of stakeholders approved release to Piccaso.
> 
> All pallets included in Picasso runtime has sufficient amount of audit.

Reviewer: ```suggestion
#### Overview of core quality gates 

Minimal set of gates:

- All changes to runtime are forward compatible.
- - If extrinsic was added, it is retained and new versions are added with larger dispatch identifiers.
- - If storage was added, it is migrated by relevant runtime code on upgrade to the new version. Storage is not upgraded in place as there are consumers which may not be able to read it after it is changed. The same relates to events.

- The runtime was upgraded on Dali Rococo Testnet and produced blocks here for several hours.

- Commit the form that validates the release as done and having passed the complete set of relevant checks on the protected main branch. 
- - You cannot force push. 
- - Each protocol on chain can stop operating as a decision of its governance without dependency on runtime upgrade.

- Reasonable representatives of stakeholders approved the release to Piccaso.

- All pallets included in Picasso runtime have been audited sufficiently.
```
Question: When you say "commit form" do you mean to commit a "release readme document"?
As I understand this is a document by the governance body and proves the release as complete and successfully checked.

### Collect release notes

If one wants to release changes to runtime, he shares merged pull requests or `done` project items linked to git commits with `release engineers`.

Only commits from main protected branch are accepted.

These can be `@ComposableFi/developers` or product owners, but can be anybody qualified.

> Release engineers indicate good receipt of ask or request more information about proposed changes. 

Reviewer: ```suggestion
Release engineers confirm proper receipt of the proposed changes or request more information.
```
Nit: Expression

### Alignment kick off

Release engineers produce `wasm` from commit of protected branch which contains all relevant commits considered for inclusion in Picasso upgrade.

After some preliminary checks and aligment not specified here, Release engineers upgrade runtime on Dali Rococo Testnet using `sudo` key. 

> Release engineers share relevant release notes, `wasm` refence and how to access runs of new runtime to stakeholder via relevant channel.

Reviewer: ```suggestion
Release engineers share relevant release notes, `wasm` references and how to access runs of new runtime to stakeholders via the relevant channels.
```

> ### Alignment kick off
> 
> Release engineers produce `wasm` from commit of protected branch which contains all relevant commits considered for inclusion in Picasso upgrade.
> 
> After some preliminary checks and aligment not specified here, Release engineers upgrade runtime on Dali Rococo Testnet using `sudo` key. 
> 
> Release engineers share relevant release notes, `wasm` refence and how to access runs of new runtime to stakeholder via relevant channel.

Reviewer: ```suggestion
### Alignment kick-off

Release engineers produce `wasm` from the commit of the protected branch, which contains all relevant commits considered for inclusion in the Picasso upgrade.

After some preliminary checks and alignment not specified here, Release engineers upgrade runtime on Dali Rococo Testnet using `sudo` key. 

Release engineers share relevant release notes, `wasm` references, and how to access runs of new runtime to stakeholders via relevant channel.
```

> ### Align
> 
> Here stakeholders do their quality gates and vote aggree with release. 
> Amount of aggreent and exact teams to aggree depend on properties of upgrade. 
> Examples, complexity of upgrade, need to hot fix.
> 
> On top of `core quality gates` next things can be considered:
> 
> `@ComposableFi/testers` runs full relevant set of integrations tests on Dali Rococo Testnet(Testnet for short).
> 
> `@ComposableFi/technical-writers` ensure that available documentation is not in conflcit with Picasso  upgrade.
> 
> `@ComposableFi/blockchain-integrations`, `@ComposableFi/bots` consider that user interface, historical data explorer and bot integrations will not negatively imact relevant integration and user experience.
> 
> `@ComposableFi/security` consider that runtime configuration(default values and included pallets) of Picasso is secure enough for relase. 
> 
> Amount and structure of alignemnt is not specified here, but some should be.

Reviewer: ```suggestion
### Align

Here stakeholders do their quality gates and vote agree with the release. 
The amount of agreement necessary and the exact teams to agree depend on the properties of the upgrade. 
Examples: complexity of upgrade, need to hot fix.

On top of `core quality gates` the following can be considered:

`@ComposableFi/testers` runs the full relevant set of integrations tests on Dali Rococo Testnet(Testnet for short).

`@ComposableFi/technical-writers` ensure that available documentation is not in conflict with the Picasso upgrade.

`@ComposableFi/blockchain-integrations`, `@ComposableFi/bots` consider that user interface, historical data explorer and bot integrations will not negatively impact relevant integration and user experience.

`@ComposableFi/security` consider that Picasso's runtime configuration(default values and included pallets) is secure enough for release. 

Amount and alignment structure are not specified here, but some should be.
```

> Here stakeholders do their quality gates and vote aggree with release. 
> Amount of aggreent and exact teams to aggree depend on properties of upgrade. 
> Examples, complexity of upgrade, need to hot fix.
> 
> On top of `core quality gates` next things can be considered:
> 
> `@ComposableFi/testers` runs full relevant set of integrations tests on Dali Rococo Testnet(Testnet for short).
> 
> `@ComposableFi/technical-writers` ensure that available documentation is not in conflcit with Picasso  upgrade.
> 
> `@ComposableFi/blockchain-integrations`, `@ComposableFi/bots` consider that user interface, historical data explorer and bot integrations will not negatively imact relevant integration and user experience.
> 
> `@ComposableFi/security` consider that runtime configuration(default values and included pallets) of Picasso is secure enough for relase. 
> 
> Amount and structure of alignemnt is not specified here, but some should be.

Reviewer: ```suggestion
Here stakeholders act on their respective quality gates and vote to agree with the release. 
The necessary agreement and the exact teams to take part, depend on the properties of the upgrade. 
I.e. the complexity of the upgrade, a need to hot-fix.

On top of `core quality gates` the following can be considered:

`@ComposableFi/testers` runs a full relevant set of integrations tests on Dali Rococo Testnet(Testnet for short).

`@ComposableFi/technical-writers` ensure that available documentation is not in conflict with the Picasso upgrade.

`@ComposableFi/blockchain-integrations`, `@ComposableFi/bots` consider that user interface, historical data explorer and bot integrations will not negatively impact relevant integration and user experience.

`@ComposableFi/security` consider that the runtime configuration(default values and included pallets) of Picasso is secure enough for release. 

Amount and structure of alignment are not specified here but should be decided.
```

### On chain upgrade operations

#### Preparing preimage

After enough aligment happend, `Release engineers` upload runtime `wasm` on chain as ***preimage***.

Collected notes, preimage hash and alignment output is send to channel with council representatives. 

#### Council 

### As SRE, we help to deploy changes to the runtime

- Given valid and credible git hash
- We ensure that runtime `wasm` with referenced code is available in `GitHub Releases`
- We collect all hashes to include in the next runtime upgrade.
- We ask for consensus from  (QA) and  (docs are up to date with runtime) and @ComposableFi/security and (UI/Frontend/Explorer) are for upgrade including relevant hashes

### Without SUDO

- @ComposableFi/sre run CI job which uploads preimage to upgrade runtime on behalf of sudo
- Sent message to `council` members about the new upgrade.

### Council collective

> - Coincil motion

Reviewer: ```suggestion
- Council motion
```

> - Coincil motion

Reviewer: ```suggestion
- Council motion
```

# Overview

Proposes flow to do runtime upgrades, likely similar to what others do.

> Expected that the reader has understood official Parity documents and explainers regarding upgrades.

Reviewer 2: Please provide references

## Constraints

- Upgrades can be done only under `Root` origin.

> - Upgrades are not immediate because runtime should be copied to the majority of para chain collators and relay validators and be enabled simultaneously on a specific block in the future.

Reviewer 2: ```suggestion
- Upgrades are not immediate because runtime should be copied to the majority of parachain collators and relay validators and be enabled simultaneously on a specific block in the future.
```

## Flow

### I want some code to be running in Picasso parachain

I make sure that:

- All code is on main protected branch of this repo.
- Code is part `dali` runtime configuration.
- Code is part of `picasso` runtime configuration.
- Relevant(see other documents regarding acceptance criteria) preliminary checks, tests, and audits passed for added code.
- I send git commit hash to #sre to ask to deploy `wasm` including that hash.

### As SRE, we help to deploy changes to the runtime

- Given valid and credible git hash
- We ensure that runtime `wasm` with referenced code is available in `GitHub Releases`
- We collect all hashes to include in the next runtime upgrade.
- We ask for consensus from @ComposableFi/testers (QA) and @ComposableFi/technical-writers (docs are up to date with runtime) and @ComposableFi/security and @ComposableFi/blockchain-integrations (UI/Frontend/Explorer) are for upgrade including relevant hashes

### With SUDO (key, multi-signature, proxy)

> @ComposableFi/sre follow Parity docs on runtime upgrade.

Reviewer 2: Please provide references

Author: fixed in commits
